src/passes/SimplifyGlobals.cpp

```diff
@@ -65,9 +80,53 @@ struct GlobalUseScanner : public WalkerPass<PostWalker<GlobalUseScanner>> {
 
   GlobalUseScanner* create() override { return new GlobalUseScanner(infos); }
 
-  void visitGlobalSet(GlobalSet* curr) { (*infos)[curr->name].written = true; }
+  void visitGlobalSet(GlobalSet* curr) { (*infos)[curr->name].written++; }
+
+  void visitGlobalGet(GlobalGet* curr) { (*infos)[curr->name].read++; }
+
+  void visitIf(If* curr) {
+    // We are looking for
+    //
+    //   if (global == X) { global = Y }
+    //
+    // Ignore an if-else, which cannot be that.
+    if (curr->ifFalse) {
+      return;
+    }
```

Comment:
The first can work. I haven't seen it in j2cl output, but it's easy to do, good idea.

The second is not valid, though. If it has some other side effect than globals, like say writing to a local, then it matters what value the global has. So we can't remove the global. For example, if might set the global from 0 to 1 the first time we get here, and all other times it increments some local.

Reply:
Actually, thinking more on the first, I'm not sure it's worth it. If we start with if-else, we may as well do if-else-if etc. nesting - but then it gets more complicated. I think maybe we can wait on that to see if there is a use case. For now this PR handles the j2cl output that exists.

Reply:
> The second is not valid, though. If it has some other side effect than globals, like say writing to a local, then it matters what value the global has. So we can't remove the global. For example, if might set the global from 0 to 1 the first time we get here, and all other times it increments some local.

Even if it has other side effects, if what it does with the specified global is only reading it, wouldn't it be possible to replace it with a constant, because it is guaranteed that there's no other reads or writes to the global?

But anyway I understand this pattern has not been observed a lot yet and you want to optimize the common case first. Just wondering why the side effects, if they are not about the global, matter.

Reply:
Oh sorry, nevermind... https://github.com/WebAssembly/binaryen/pull/4070#discussion_r686108438 explains that.